File: streams/examples/06-health.js
import { Queen, Stream } from 'queen-mq'
const url = process.env.QUEEN_URL || 'http://localhost:6632'

const q = new Queen({ url, handleSignals: false })

const QUERY_ID = 'patient-heart-rate-aggregator'
const Queue = 'patient-heart-rate' 

await q.queue(Queue).delete()
await q.queue(Queue).create()

async function patientHeartRateProducer() {
    console.log('Starting patient heart rate producer')
    while (true) {
        for (let i = 0; i < 10; i++) {
            await q.queue(Queue).partition(`patient-${i}`).push({
                data: {
                    patientId: `patient-${i}`,
                    heartRate: Math.floor(Math.random() * 100)
                }
            })
        }
        await new Promise(resolve => setTimeout(resolve, 1))
    }
}

patientHeartRateProducer()


const handle = await Stream
  .from(q.queue(Queue))
  .windowTumbling({ seconds: 5 })
  .reduce(
    (acc, m) => {
      acc.count += 1
      acc.sum   += m.heartRate
      acc.min    = acc.min === null ? m.heartRate : Math.min(acc.min, m.heartRate)
      acc.max    = acc.max === null ? m.heartRate : Math.max(acc.max, m.heartRate)
      //acc.messages.push(m)         // collect every message that fell in this window
      return acc
    },
    { count: 0, sum: 0, min: null, max: null, messages: [] }
  )  
  //.aggregate({
  //  count: () => 1,
  //  sum:   m => m.heartRate,
  //  min:   m => m.heartRate,
  //  max:   m => m.heartRate,
  //  avg:   m => m.heartRate
  //})
  // The aggregate emit value is the {count, sum, min, max} object — we
  // attach a marker for visibility. .map runs as a post-reducer stage.
  //.map(agg => ({ ...agg, kind: 'window-emit' }))
  //.to(q.queue(SINK_QUEUE))
  .foreach(async (window, ctx) => {
    // window === { count, sum, min, max, avg, emittedAt }
    console.log('closed window:', window, ctx)
  })  
  .run({
    queryId: QUERY_ID,
    url,
    batchSize: 50,
    maxPartitions: 10,
    reset: true
  })

File: clients/client-py/examples/06_health.py
"""
Example 06 — Patient health-rate stream. 1:1 port of 06-health.js.

Pushes simulated heart-rate readings across N patients and runs a tumbling
windowed aggregation. Equivalent in shape to 00_demo.py.

Run:
    QUEEN_URL=http://localhost:6632 python examples/06_health.py
"""

from __future__ import annotations

import asyncio
import os
import random
import time

from queen import Queen, Stream


URL = os.environ.get("QUEEN_URL", "http://localhost:6632")
TAG = f"{int(time.time() * 1000):x}"
SOURCE_QUEUE = f"streams_health_hr_{TAG}"
QUERY_ID = f"streams.health.{TAG}"
PATIENTS = [f"patient-{i}" for i in range(10)]


async def main():
    q = Queen(url=URL)
    try:
        await q.queue(SOURCE_QUEUE).create()

        async def producer():
            print("Starting patient heart rate producer")
            for _ in range(60):
                await asyncio.gather(*[
                    q.queue(SOURCE_QUEUE).partition(p).push([
                        {"data": {"patient": p, "heartRate": random.randint(40, 100)}}
                    ])
                    for p in PATIENTS
                ])
                await asyncio.sleep(1)

        prod_task = asyncio.create_task(producer())

        def on_window(agg, ctx):
            print("closed window:", agg, ctx)

        handle = await (
            Stream.from_(q.queue(SOURCE_QUEUE))
            .window_tumbling(seconds=10, idle_flush_ms=2000)
            .aggregate({
                "count": lambda m: 1,
                "sum":   lambda m: m["heartRate"],
                "min":   lambda m: m["heartRate"],
                "max":   lambda m: m["heartRate"],
                "avg":   lambda m: m["heartRate"],
            })
            .map(lambda agg, ctx: ({**agg, "kind": "window-emit"}, on_window(agg, ctx))[0])
            .run(query_id=QUERY_ID, url=URL, batch_size=50, max_partitions=10, reset=True)
        )

        await prod_task
        await asyncio.sleep(15)
        await handle.stop()
        print("Stats:", handle.metrics())
    finally:
        await q.close()


if __name__ == "__main__":
    asyncio.run(main())
